src/matrixrtc/NewMembershipManager.ts

```diff
@@ -0,0 +1,945 @@
+                    return {
+                        insert: [
+                            {
+                                ts: Date.now(),
+                                type: MembershipActionType.SendFirstDelayedEvent,
+                            },
+                        ],
+                    };
+                }
+                const updateNetwork = this.actionUpdateFromNetworkErrorRetry(e, type);
+                if (updateNetwork) return updateNetwork;
+                const updateLimit = this.actionUpdateFromRateLimitError(e, "sendDelayedStateEvent", type);
+                if (updateLimit) return updateLimit;
+
+                // log and fall through
+                if (this.isUnsupportedDelayedEndpoint(e)) {
+                    logger.info("Not using delayed event because the endpoint is not supported");
+                } else {
+                    logger.info("Not using delayed event because: " + e);
+                }
+                // On any other error we fall back to not using delayed events and send the join state event immediately
+                return createInsertActionUpdate(MembershipActionType.SendJoinEvent);
+            });
+    }
+
+    private async cancelKnownDelayIdBeforeSendFirstDelayedEvent(
+        type: MembershipActionType,
+        delayId: string,
+    ): Promise<ActionUpdate> {
+        // Remove all running updates and restarts
+        const resetActionUpdate = { replace: [] };
+        return await this.client
+            ._unstable_updateDelayedEvent(delayId, UpdateDelayedEventAction.Cancel)
+            .then(() => {
+                this.state.delayId = undefined;
+                this.resetRateLimitCounter(MembershipActionType.SendFirstDelayedEvent);
+                return createReplaceActionUpdate(MembershipActionType.SendFirstDelayedEvent);
+            })
+            .catch((e) => {
+                const updateLimit = this.actionUpdateFromRateLimitError(e, "updateDelayedEvent", type);
+                if (updateLimit) return { ...resetActionUpdate, ...updateLimit };
+                const updateNetwork = this.actionUpdateFromNetworkErrorRetry(e, type);
+                if (updateNetwork) return { ...resetActionUpdate, ...updateNetwork };
+
+                if (this.isNotFoundError(e)) {
+                    // If we get a M_NOT_FOUND we know that the delayed event got already removed.
+                    // This means we are good and can set it to undefined and run this again.
+                    this.state.delayId = undefined;
+                    return createReplaceActionUpdate(MembershipActionType.SendFirstDelayedEvent);
+                }
+                if (this.isUnsupportedDelayedEndpoint(e)) {
+                    return createReplaceActionUpdate(MembershipActionType.SendJoinEvent);
+                }
+
+                // This becomes an unhandle-able error case since sth is signifciantly off if we dont hit any of the above cases
```

Comment:
```suggestion
                // This becomes an unrecoverable error case since something is significantly off if we don't hit any of the above cases
```